crates/ra_hir/src/generics.rs

```diff
@@ -45,18 +45,26 @@ impl GenericParams {
     ) -> Arc<GenericParams> {
         let mut generics = GenericParams::default();
         let parent = match def {
-            GenericDef::Function(it) => it.impl_block(db),
-            GenericDef::TypeAlias(it) => it.impl_block(db),
+            // FIXME abstract over containers (trait/impl)
+            GenericDef::Function(it) => it
+                .impl_block(db)
+                .map(GenericDef::from)
+                .or_else(|| it.parent_trait(db).map(GenericDef::from)),
```

Comment:
Yeah, in this case I like the separate steps more, since it keeps a bit more symmetry ;) In any case, this handling of impl vs. trait parents shouldn't even happen here.